docs/proposals/015-bindless-t.md

```diff
@@ -0,0 +1,160 @@
+SP #015 - `Bindless<T>` type
+==============
+
+## Status
+
+Author: Yong He
+
+Status: Design review.
+
+Implementation:
+
+Reviewed by:
+
+## Background
+
+Textures, sampler states and buffers are typically passed to shader as opaque handles whose size and storage address is undefined. These handles are communicated to the GPU via "bind states" that are modified with host-side APIs. Because the handle has unknown size, it is not possible to read, copy or construct such a handle from the shader code, and it is not possible to store the handle in buffer memory. This makes both host code and shader code difficult to write and prevents more flexible encapsulation or clean object-oriented designs.
+
+With the recent advancement in hardware capabilities, a lot of modern graphics systems are adopting a "bindless" parameter passing idiom, where all resource handles are passed to the shader in a single global array, and all remaining references to texture, buffers or sampler states are represented as a single integer index into the array. This allows the shader code to workaround the restrictions around the opaque handle types.
+
+Direct3D Shader Model 6.6 introduces the "Dynamic Resources" capability, which further simplifies the way to write bindless shader code by removing the need to even declare the global array.
+
+We believe that graphics developers will greatly benefit from a system defined programming model for the bindless parameter passing idom that is versatile and cross-platform, which will provide a consistent interface so that different shader libraries using the bindless pattern can interop with each other without barriers.
+
+## Proposed Approach
+
+We introduce a `Bindless<T>` type that is defined as:
+```
+struct Bindless<T> : IComparable<Bindless<T>>
+    where T : IOpaqueHandle
+{
+    __init(uint64_t value);
+}
+```
+Where `IOpaqueHandle` is an interface that is implemented by all texture, buffer and sampler state types:
+
+```slang
+enum ResourceKind
+{
+    Unknown, Texture, ConstantBuffer, StorageBuffer, Sampler
+}
+interface IOpaqueHandle
+{
+    static ResourceKind getResourceKind();
```

Comment:
Should this effectively be an associated constant?

Reply:
yes that's better, fixed.